Feature - Enable Mocha glob pattern

ejfrancis opened the pull request:
see #554 

Given a `--files` flag that is a glob, any files matching that glob will be added to the mocha test. Files in the `--path` directory are still added

```
chimp --ddp=http://localhost:3005 --mocha --path=tests --files=./client/**/*.test.js
```

@samhatoum let me know if you'd prefer a different public API but this gets it working

ejfrancis commented:
I'm not sure how I would go about testing this

samhatoum commented:
You can spy on Mocha.addFiles, supply a glob pattern to a new MochaWrapper and assert that addFiles was called.

You might want to use the new wallaby/mocha/testdouble test rig I just created. I'm about to merge that into master as soon as the build passes. See here: https://github.com/xolvio/chimp/pull/551

samhatoum commented:
The new test rig is MUCH faster and MUCH easier to use. Testdouble.js is awesome

ejfrancis commented:
alright I'll give it a shot

ejfrancis commented:
I'm having a hard time figuring out the right way to do this with testdouble

I've got this so far

```javascript
var Mocha = require('Mocha');
var glob = require('glob');
var path = require('path');
var fs = require('fs');

var filesGlobParam = './src/**/*.spec.js';
var testDir = './tests/';
var testDirFiles = ['fileA.js', 'fileB.js'];


describe('mocha-wrapper', function () {
  before(function () {
    td.replace('Mocha', td.object(Mocha));
    td.replace('fs', td.object(fs));
    td.replace('path', td.object(path));
    td.replace('glob', td.object(glob));
    td.replace('exit', td.function('exit'));
    td.replace('glob', td.object({ sync: (path) => path }));   // just return the path given
  });
  describe('add files', function () {
    it('adds files inside testDir', function () {
      process.env['chimp.path'] = testDir;
      td.when(path.join(testDir, '**')).thenReturn(testDir);
      td.when(glob.sync(testDir)).thenReturn(testDirFiles);

      require('./mocha-wrapper');
      td.verify(mocha.addFile was called once with 'fileA.js' and once with 'fileB.js');
    });
  });
});

```

I think you can see what I'm trying to do. Assert that the files inside `process.env['chimp.path']` are run through `mocha.addFile()`. I'm not trying to test the `--files` flag yet, just the `--path` flag.

`mocha-wrapper` doesn't export any constructor so the only way to evaluate it is to `require()` it. After that, I want to verify that `mocha.addFiles()` was called once with `fileA.js` and once with `fileB.js`. I've made the `Mocha` constructor a `td.object()`, but how can I get a reference to the `addFiles()` function on new instances of `Mocha()` created in `mocha-wrapper` and assert that it was called with the proper arguments across multiple calls, since `mocha.addFile()` will be called a few times?

samhatoum commented:
Just did a bit of research and found this issue:
https://github.com/testdouble/testdouble.js/issues/166

It's a little late for my brain at 3AM. I'll try again soon but you might have better luck in the meantime

ejfrancis commented:
it looks like that would work for this use case if `mocha-wrapper` returned a constructor, but since it doesn't I have to just `require()` it to make it evaluate and that doesn't seem to play too well with testdouble. I'm thinking of splitting the `mocha-wrapper` into `mocha-wrapper` (constructor) and `mocha-wrapper-instance` (equivalent of `mocha-wrapper` right now) so that I can import it into tests

samhatoum commented:
I agree. It's better design also as we can make the wrapper more modular.

ejfrancis commented:
alright new changes:

* `mocha-wrapper.js` now exports a constructor that just runs the code that was in the file before
* `mocha-wrapper-instance.js` is now the equivalent of the old `mocha-wrapper.js`, it just imports the constructor and creates a new instance
* added unit tests in `mocha-wrapper-spec.js` that verify the `--path` and `--files` flags work
* changed `mocha.js` to create a new child process of `mocha-wrapper-instance` instead of pointing to `mocha-wrapper`

looks like the tests failed on circle, now sure why since they pass on my local

ejfrancis commented on src/lib/mocha/mocha-wrapper.js:
```diff
@@ -10,41 +9,57 @@ var Mocha = require('mocha'),
-  var mochaConfig = JSON.parse(process.env.mochaConfig);
-mochaConfig.ui = 'fiberized-bdd-ui';
+class MochaWrapper {
+  constructor() {
```
everything inside this constructor was just copy/pasted from the contents of the old file, nothing's been changed but adding the section for `process.env['chimp.files']`

ejfrancis commented on src/lib/mocha/mocha.js:
```diff
@@ -70,7 +70,7 @@ Mocha.prototype.start = function (callback) {
   });
 
-  self.child = cp.fork(path.join(__dirname, 'mocha-wrapper.js'), _.union(
+  self.child = cp.fork(path.join(__dirname, 'mocha-wrapper-instance.js'), _.union(
```
this looked like the only place dependent on the old `mocha-wrapper.js` but let me know if I missed somewhere else

samhatoum replied:
no, that's it. You're right. I wish there was a nice way to do it than to have a file, but this is the implication of dealing with processes I guess.

ejfrancis commented:
nice fixed the tests :) I guess it didn't like this

```
import glob from 'glob';

td.replace('glob', td.object(glob));
```

so i had to mock out glob just with an object `{ sync: () = {} }`

ejfrancis commented on src/lib/mocha/mocha-wrapper.js:
```diff
@@ -10,41 +9,57 @@ var Mocha = require('mocha'),
+    var mocha = new Mocha(mochaConfig);
+
+    mocha.addFile(path.join(path.resolve(__dirname, path.join('mocha-helper.js'))));
+    if (process.argv.length > 3) {
```
I don't understand this if block so I didn't add a test case around it, so this is the only `mocha.addFile()` call that isn't tested

lgandecki replied:
this allows you to pass test files in a list.. test/test1.js test/test2.js test/test3.js . If you use circleci parallelization this is a must, since that's what circle provides to it's parallelized build machines

Looking at it now - I definitely agree, this could be made much more cleaner ;)

ejfrancis replied:
k just added a test for this case

samhatoum commented:
This all looks awesome. I just need to test it all locally
